src/sparse_blas/backends/mklcpu/mklcpu_operations.cpp

```diff
@@ -0,0 +1,209 @@
+    const fpType /*beta*/, sycl::buffer<fpType, 1>& /*C*/, const std::int64_t /*ldc*/) {
+    throw unimplemented("SPARSE_BLAS", "gemm");
+}
+
+template <typename fpType>
+std::enable_if_t<detail::is_fp_supported_v<fpType>, sycl::event> gemm(
+    sycl::queue& /*queue*/, layout /*dense_matrix_layout*/, transpose /*transpose_A*/,
+    transpose /*transpose_B*/, const fpType /*alpha*/, detail::matrix_handle* /*A_handle*/,
+    const fpType* /*B*/, const std::int64_t /*columns*/, const std::int64_t /*ldb*/,
+    const fpType /*beta*/, const fpType* /*C*/, const std::int64_t /*ldc*/,
+    const std::vector<sycl::event>& /*dependencies*/) {
+    throw unimplemented("SPARSE_BLAS", "gemm");
+}
+
+#define INSTANTIATE_OPERATIONS(FP_TYPE)                                                           \
+    template std::enable_if_t<detail::is_fp_supported_v<FP_TYPE>> gemv(                           \
+        sycl::queue& queue, transpose transpose_val, const FP_TYPE alpha,                         \
+        detail::matrix_handle* A_handle, sycl::buffer<FP_TYPE, 1>& x, const FP_TYPE beta,         \
+        sycl::buffer<FP_TYPE, 1>& y);                                                             \
+    template std::enable_if_t<detail::is_fp_supported_v<FP_TYPE>, sycl::event> gemv(              \
+        sycl::queue& queue, transpose transpose_val, const FP_TYPE alpha,                         \
+        detail::matrix_handle* A_handle, const FP_TYPE* x, const FP_TYPE beta, const FP_TYPE* y,  \
+        const std::vector<sycl::event>& dependencies);                                            \
+    template std::enable_if_t<detail::is_fp_supported_v<FP_TYPE>> gemvdot(                        \
+        sycl::queue& queue, transpose transpose_val, FP_TYPE alpha,                               \
+        detail::matrix_handle* A_handle, sycl::buffer<FP_TYPE, 1>& x, FP_TYPE beta,               \
+        sycl::buffer<FP_TYPE, 1>& y, sycl::buffer<FP_TYPE, 1>& d);                                \
+    template std::enable_if_t<detail::is_fp_supported_v<FP_TYPE>, sycl::event> gemvdot(           \
+        sycl::queue& queue, transpose transpose_val, FP_TYPE alpha,                               \
+        detail::matrix_handle* A_handle, FP_TYPE* x, FP_TYPE beta, FP_TYPE* y, FP_TYPE* d,        \
+        const std::vector<sycl::event>& dependencies);                                            \
+    template std::enable_if_t<detail::is_fp_supported_v<FP_TYPE>> symv(                           \
+        sycl::queue& queue, uplo uplo_val, FP_TYPE alpha, detail::matrix_handle* A_handle,        \
+        sycl::buffer<FP_TYPE, 1>& x, FP_TYPE beta, sycl::buffer<FP_TYPE, 1>& y);                  \
+    template std::enable_if_t<detail::is_fp_supported_v<FP_TYPE>, sycl::event> symv(              \
+        sycl::queue& queue, uplo uplo_val, FP_TYPE alpha, detail::matrix_handle* A_handle,        \
```

Comment:
This feels super fragile. It's hard to write this up, it's hard to review this. If there's some obvious better way to do this, then great, otherwise, let me know and I'll go in and review this a lot more closely.

Reply:
I do know of one way to do this differently. We can recompile the same file differently for different data types and/or integer types as applicable into differently named object files and link those multiple object files into the library.
e.g.
```bash
icpx -fsycl -DFP_TYPE=float                -c mklcpu_operations.cpp -o mklcpu_operations_r_f.o
icpx -fsycl -DFP_TYPE=double               -c mklcpu_operations.cpp -o mklcpu_operations_r_d.o
icpx -fsycl -DFP_TYPE=std::complex<float>  -c mklcpu_operations.cpp -o mklcpu_operations_c_f.o
icpx -fsycl -DFP_TYPE=std::complex<double> -c mklcpu_operations.cpp -o mklcpu_operations_c_d.o
```
We can think about whether this route is worth exploring or not.

Reply:
> cannot be changed without a lot of duplication (like the blas domain) or a different build system because the implementations depend on FP_SUFFIX and INT_SUFFIX to call the right function in the function table

I see what you mean, the function implementation calls functions named `set_csr_data##FP_SUFFIX##INT_SUFFIX`, so we either need the macros, or we need to expand those out with code duplication. This is fine then. Let's leave it as it is.